src/nlv_object.h

```diff
@@ -48,36 +26,58 @@ class NLVObjectBasePtr
   friend class NLVObject;
 
 
-class NLVObjectBase : public Nan::ObjectWrap
+class NLVObjectBase : public Nan::ObjectWrap, public std::enable_shared_from_this<NLVObjectBase>
 {
 public:
+  void AddToParent(NLVObjectBase* parent) {
+    parent->PushChild(shared_from_this());
+  }
+  
+  void ClearChildren() {
+    for (auto& ptr: children_) {
+      if (auto object = ptr.lock()) {
+        object->ClearHandle();
+        object->ClearChildren();
+      }
+    }
+    children_.clear();
+  }
+
   virtual void ClearHandle() = 0;
-  virtual void ClearChildren() = 0;
-  virtual void SetParentReference(NLVObjectBasePtr *parentReference) = 0;
 
-  std::vector<NLVObjectBasePtr*> children_;
+  void PushChild(const std::shared_ptr<NLVObjectBase>& child) {
+    children_.emplace_front(child);
+  }
+  std::forward_list<std::weak_ptr<NLVObjectBase>> children_;
```

Comment:
I don't see the case for using a `std::forward_list` here over a `std::vector`, other than you chose to `emplace_front` above. We're working with a collection that is not changing much, and certainly isn't going to need random access, so a `std::vector` should be preferred, and is likely more performant